harness/asyncHelpers.js

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+description: |
+    A collection of assertion and wrapper functions for testing asynchronous built-ins.
+defines: [asyncTest]
+---*/
+
+function asyncTest(testFunc) {
+  if (!Object.hasOwn(globalThis, "$DONE")) {
+    throw new Test262Error("asyncTest called without async flag");
+  }
+  if (typeof testFunc !== "function") {
+    $DONE(new Test262Error("asyncTest called with non-function argument"));
+    return Promise.resolve();
+  }
+  try {
+    return testFunc().then(
+      function () {
+        $DONE();
+      },
+      function (error) {
+        $DONE(error);
+      }
+    );
+  } catch (syncError) {
+    $DONE(syncError);
+    return Promise.resolve();
```

Comment:
why not make the containing function `async` rather than needing to always return a Promise?

Reply:
The goal is to ensure that if it's called without the async flag, it will still execute to throw the error on line 11 - I believe top-level calls to async functions might not execute in most runners before the implementation returns, leading to silent failures if an author omits the 'async' flag. But I could well be wrong!

Reply:
You're correct, but i'd expect that to be handled by the unhandled rejection handler, rather than a confusing sync exception from a promise-returning function (ie, zalgo)

Reply:
The RFC itself [makes the case](https://github.com/tc39/test262/pull/3724/files#diff-d4e487dcd72bb7cd7462c9709f50abd806977f296306b5f4eb77e3fc011ede51R112) for summoning Zalgo here, and test262 runners either don't handle unhandled rejections or handle them without clear messages. Engine262 and JSC both pass a test with an unhandled rejection; v8 will fail it with no output. Node gives an UnhandledPromiseRejection, as does SpiderMonkey. Because of these reasons, I think it's actually worth the top-level Zalgo - hopefully we only would expect this to be called top level. I'd argue it's at least moderately easier to review and catch than the possible mistakes that arise from not having it.